rate-limiter/src/token_bucket.rs

```diff
@@ -44,12 +105,16 @@ impl TokenBucket {
 
     fn update_units(&mut self, now: Instant) -> usize {
         let time_since_last_update = now.duration_since(self.last_update);
-        let new_units = time_since_last_update
+        let mut new_units = time_since_last_update
             .as_micros()
             .saturating_mul(self.rate_per_second as u128)
             .saturating_div(1_000_000)
             .try_into()
             .unwrap_or(usize::MAX);
+        let used_new_units = min(new_units, self.requested);
+        new_units -= used_new_units;
+        self.requested -= used_new_units;
```

Comment:
Do you need the `requested` field? Wouldn't `available` suffice? You'd call `self.update_units(now)` at the entry to `rate_limit`. It'd calculate how much time has passed since `self.last_update` and add appropriate amount of `new_units` to `self.available`. If you used `AtomicU64` then you could use `fetch_update` – `self.available.fetch_update(..., ..., available.saturating_sub(requested))`. This will return `available` before the call so you can now (lock-free) return to the caller `diff = requested - available` - covering the case where `requested > available` in which case the caller is asked to `sleep(diff)`. 

If you manage to work with a single `AtomicU64` I think it's possible to get rid of the `lock` on the call-side.

Reply:
Note: I think I am using a different definition of `lock-free` here. Did you mean something more like `wait-free` (progress for every thread...per definition on wikipedia)? `fetch_update` can block theoretically...but this not important here.

It is possible to make this algorithm `wait-free`. Main concern here was actually `starvation` of `good` peers by someone sending huge amount of data. With this mutex it behaves like a concurrent queue - `bad` actor can't fully starve `good` nodes, but can slow down everything significantly. That's why I've decided to switch it back to `per-node` rate-limiting version and tried to figure out if there is some other solution of this problem. One problem with per-node approach is that it is harder to choose good enough constants that scale with number of peers (with same hardware) - it is easier to find value like: my hardware can handle 8Mib/s of query bandwidth.

Reply:
By _lock-free_ I meant the very naive version – simply don't use `.lock()` calls to `Mutex`. 

Thanks for the comments and changes. I'll study them and learn :)